app/jobs/publish_feed_job.rb

```diff
@@ -18,30 +18,59 @@ def perform(podcast, pub_item)
     podcast.feeds.each { |feed| publish_apple(podcast, feed) }
     podcast.feeds.each { |feed| publish_rss(podcast, feed) }
     PublishingPipelineState.complete!(podcast)
+  rescue Apple::AssetStateTimeoutError => e
+    handle_apple_timeout_error(podcast, e)
   rescue => e
-    PublishingPipelineState.error!(podcast)
-    Rails.logger.error("Error publishing podcast", {podcast_id: podcast.id, error: e.message, backtrace: e.backtrace.join("\n")})
-    raise e
+    handle_error(podcast, e)
   ensure
     PublishingPipelineState.settle_remaining!(podcast)
   end
 
+  def handle_apple_timeout_error(podcast, error, raise_error: true)
+    PublishingPipelineState.error!(podcast)
+    Rails.logger.error("Asset processing timeout", {podcast_id: podcast.id, error: error.message, backtrace: error&.backtrace&.join("\n")})
+    raise error
+  end
+
+  def handle_error(podcast, error)
+    PublishingPipelineState.error!(podcast)
+    # Two error log lines here.
+    # 1) the error message and backtrace:
+    Rails.logger.error("Error publishing podcast", {podcast_id: podcast.id, error: error.message, backtrace: error&.backtrace&.join("\n")})
+    # 2) The second is from the job handler, which logs an error when this excetion is raised:
+    raise error
+  end
+
   def publish_apple(podcast, feed)
     return unless feed.publish_to_apple?
-    res = PublishAppleJob.perform_now(feed.apple_config)
-    PublishingPipelineState.publish_apple!(podcast)
-    res
-  rescue => e
-    NewRelic::Agent.notice_error(e)
-    res = PublishingPipelineState.error_apple!(podcast)
-    raise e if feed.apple_config.sync_blocks_rss?
-    res
+
+    begin
+      res = PublishAppleJob.do_perform(podcast.apple_config)
+      PublishingPipelineState.publish_apple!(podcast)
+      res
+    rescue => e
+      handle_apple_error(podcast, e)
```

Comment:
So the overall flow through these 3 rescue blocks is something like:

1. We call `publish_apple` for the feeds
2. If PublishAppleJob errors, we rescue here.  This sets an `error_apple!` pipeline state, alerts NR, and raises.
3. We rescue again up in `def perform` - which either `handle_apple_timeout_error` or `handle_error`s it
4. That log.errors and sets a 2nd `error!` pipeline state

So we're setting 2 pipeline states per failure, NewRelic-noticing in the inner-error-handler, and rails.logger-ing in the outer-error handler.  Did I get that right?

Reply:
Yup, the gist is that there are two layers of error handling. So that we can set two errors. The outer error handler represents the terminal error state. And the inner errors (error_rss, error_apple) represent transient moves through the publishing pipeline state.

Reply:
Good catch, though to bring up the NR notice -- we're missing the `sync_blocks_rss` handling here